Gromacs function for Iluvatar

Bhatt21 opened the pull request:
Gromacs example
The user can specify URL(tpr file download URL),nsteps, and modes.
This has 3 modes, CPU,GPU, and hybrid(default)
If set to either, will offload as much tasks as possible to CPU/GPU (see comments, have linked documentation)

This PR also contains a Docker image for a normal CPU image.

I also wanted to add a GPU image, I did in fact make one using [this](https://github.com/COS-IN/iluvatar-faas/blob/master/src/load/functions/python3/Dockerfile.gpu)
But the problem is that the final image is enormous.
I would need some help in reducing its size as I don't have much experience with nvidia .

aFuerst commented on src/load/functions/python3/functions/gromacs/main.py:
```diff
@@ -0,0 +1,88 @@
+
+    # From documentation
+    # -nb Used to set where to execute the short-range non-bonded interactions. 
+    # Can be set to “auto,” “cpu” or “gpu.” Defaults to “auto,” which uses a 
```
You have `auto` (default value of gromacs I assume?) here in your documentation, but your default value is `hybrid`. Is that intentional?

Bhatt21 replied:
thanks
hybrid -> auto  done

aFuerst commented on src/load/functions/python3/functions/gromacs/main.py:
```diff
@@ -0,0 +1,88 @@
+    end_time = time.time()
+
+    return {
+        "cold": was_cold,
```
You aren't returning/exporting any results from the simulation?

Bhatt21 replied:
returning output

aFuerst commented:
> GPU images are generally large because of the aircraft-carrier sized cuda runtime. If it still can be launched by Iluvatar, then we are fine!

Yup, it's 99% CUDA at fault for being huge. What Nvidia dependencies you might need can vary wildly based on application. Our Dockerfile.gpu just copies basically every cuda .so file because ML projects use most everything and nailing down exactly which files and versions was a hassle

You can test your image manually via docker by attaching a [local GPU via a flag](https://docs.docker.com/reference/cli/docker/container/run/#gpus). This is essentially what we do anyway. See if gromacs w/gpu accel works on the giant image, see if it works on your skinny image. Then if you care enough, you could copy just the minimum [like we used to do](https://github.com/COS-IN/iluvatar-faas/blob/master/src/load/functions/python3/Dockerfile.gpu#L14)

aFuerst commented:
@Bhatt21 have you been able to work on this?

Bhatt21 commented:
> @Bhatt21 have you been able to work on this?

Hi @aFuerst  I will fix and close this soon, was busy with some other commitment

Bhatt21 commented:
Hi @aFuerst 
have added gpu image and addressed your comments, please have a look
just for comparison some benchmarks
dataset: benchMEM
nsteps 1000

for
CPU 2, mem 2048, GPU 0 -> time taken ~17 seconds
CPU 2 mem 2048 GPU 1 -> time taken ~6 seconds

dataset: benchMEM
nsteps 5000
CPU 2, mem 2048, GPU 0 -> time taken ~70 seconds
CPU 2 mem 2048 GPU 1 -> time taken ~20 seconds

aFuerst reviewed (approved):
Nice work, LGTM!
